File: python-agent/kimo-engine-preview/assistant/intent.ts
import type { Message } from './store'

/**
 * تحليل نية المستخدم من نص طلبه — تصنيف عربي مختصر يُحقن في سياق الوكيل
 * ليساعده على اختيار المسار الصحيح والتحسن المستمر في فهم الطلبات.
 */
export function analyzeIntent(text: string): string {
  const t = String(text ?? '').trim()
  if (!t) return ''
  if (/احذف|امسح|حذف|إلغاء|ألغ|شيل|احدف/.test(t) && !/إلغاء آخر|تراجع/.test(t)) return 'حذف'
  if (/تراجع|إلغاء آخر عملية|اعكس آخر/.test(t)) return 'تراجع'
  if (/أنشئ|أنشأ|أضف|سجّل|سجل|ادخل|أدخل|اعمل مشروع|اعمل جديد|انشئ/.test(t)) return 'إنشاء/إضافة'
  if (/عدّل|عدل|حدّث|حددث|غيّر|غير|صحّح|صحح|استبدل/.test(t)) return 'تعديل'
  if (/استيراد|اقرأ الملف|اقرأ ملف|رفع ملف|استورد|حول الملف/.test(t)) return 'استيراد ملفات'
  if (/ملف|تقرير|جدول إكسل|إكسل|Excel|Word|PDF|ورقة|ورود/.test(t) && /ولّد|ولد|جهز|اعمل|أنشئ|اكتب|أكتب/.test(t)) return 'توليد ملف'
  if (/ملخص|خلاصة|تحليل|إحصاء|احسب|حسب|راجع|قارن|قارن بين|مؤشرات/.test(t)) return 'تحليل/ملخص'
  if (/ابحث|بحث|دور|وين|أين|استعلام|جيب|اعرض|أعرض|عرض كل|قائمة/.test(t)) return 'بحث/استعلام'
  if (/مرحبا|هلا|اهلا|كيف حالك|شكرا|شكرا لك|تصفح|ممكن تساعد/.test(t)) return 'محادثة عامة'
  return 'استعلام/مساعدة'
}

/** ملخص سياق المحادثة الحديثة: آخر ما كان عليه المستخدم (بدون رموز تقنية). */
export function buildContextSummary(msgs: Message[], maxUsers = 3): string {
  const users = msgs.filter((m) => m.role === 'user' && m.content && m.content.trim()).slice(-maxUsers)
  if (!users.length) return ''
  const parts = users.map((m) => {
    const c = m.content.trim()
    return c.length > 80 ? `${c.slice(0, 80)}…` : c
  })
  return `سياق المحادثة الحديثة: ${parts.join(' ← ')}`
}

File: kimo/intent.py
"""Intent analysis + conversation context summary (mirrors ``assistant/intent.ts``)."""

from __future__ import annotations

import re
from typing import Optional

_INTENT_PATTERNS: list[tuple[str, re.Pattern[str]]] = [
    ("create", re.compile(r"(أنشئ|انشئ|أضف|اضف|سجّل|سجل|استورد|استيراد|ولّد|ولد|إنشاء)")),
    ("update", re.compile(r"(عدّل|عدل|حدّث|حدث|غيّر|غير|تعديل)")),
    ("delete", re.compile(r"(احذف|حذف|إزالة|ازالة|شيل|احذفهم)")),
    ("read", re.compile(r"(اعرض|أظهر|اظهر|اقرأ|استعرض|ابحث|اعثر|كم|عدد|إجمالي|ملخص|جدول)")),
    ("ask", re.compile(r"(\?$|؟$|ماذا|كيف|لماذا|هل|ما هو|ما هي)")),
]


def analyze_intent(text: str) -> Optional[str]:
    for label, pat in _INTENT_PATTERNS:
        if pat.search(text or ""):
            return label
    return None


def build_context_summary(messages: list) -> str:
    """Summarise recent turns into a short context line for the system prompt."""
    recent = [m for m in messages if getattr(m, "role", None) in ("user", "assistant")][-6:]
    if not recent:
        return ""
    parts = []
    for m in recent:
        role = "المستخدم" if m.role == "user" else "المساعد"
        content = m.text if hasattr(m, "text") else str(getattr(m, "content", ""))
        parts.append(f"{role}: {content[:120]}")
    return "آخر المحادثة:\n" + "\n".join(parts)
